File: algorithms/strings/longest-palindromic-substring/java/LongestPalindromeSubarray.java
public class LongestPalindromeSubarray {

    public static int longestPalindromeSubarray(int[] arr) {
        int n = arr.length;
        if (n == 0) return 0;

        int maxLen = 1;
        for (int i = 0; i < n; i++) {
            int odd = expand(arr, i, i);
            int even = expand(arr, i, i + 1);
            maxLen = Math.max(maxLen, Math.max(odd, even));
        }
        return maxLen;
    }

    private static int expand(int[] arr, int l, int r) {
        while (l >= 0 && r < arr.length && arr[l] == arr[r]) {
            l--;
            r++;
        }
        return r - l - 1;
    }
}

File: algorithms/strings/longest-palindromic-substring/scala/LongestPalindromeSubarray.scala
object LongestPalindromeSubarray {

  def longestPalindromeSubarray(arr: Array[Int]): Int = {
    val n = arr.length
    if (n == 0) return 0

    def expand(l: Int, r: Int): Int = {
      var left = l
      var right = r
      while (left >= 0 && right < n && arr(left) == arr(right)) {
        left -= 1
        right += 1
      }
      right - left - 1
    }

    var maxLen = 1
    for (i <- 0 until n) {
      val odd = expand(i, i)
      val even = expand(i, i + 1)
      maxLen = math.max(maxLen, math.max(odd, even))
    }
    maxLen
  }
}
